### PR TITLE
fix: ApplicationSet webhooks should use stricter comparison of repository URL

### LINKED ISSUE #20485: ApplicationSet webhooks should use stricter comparison of repository URL
Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `argocd version`.

**Describe the bug**

ApplicationSet webhook can trigger refresh of unwanted ApplicationSet, due to regexp that is not enough strict at the end, in my opinion, it is missing a `$` at end of this regexp :

https://github.com/argoproj/argo-cd/blob/ac69f5223cd0fc42fcc638b340a3703addda3812/applicationset/webhook/webhook.go#L225

**To Reproduce**

Create two ApplicationSet where URL are close : 
```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
....
spec:
  generators:
  - git:
      files:
      - path: envs/**/config.json
      repoURL: https://github.com/dacofr/hello-appset.git
---
apiVersion: argoproj.io/v1alpha1
kind: ApplicationSet
....
spec:
  generators:
  - git:
      files:
      - path: envs/**/config.json
      repoURL: https://github.com/dacofr/hello-appset-copy.git
```

Trigger a webhook (included minimal payload from GitHub)
```
curl --request POST \
  --url http://localhost:7000/api/webhook \
  --header 'Content-Type: application/json' \
  --header 'X-GitHub-Event: push' \
  --data '{
    "ref": "refs/heads/main",
    "repository": {
      "html_url": "https://github.com/dacofr/hello-appset",
      "default_branch": "main"
    }
}'
```

**Logs**

```
INFO[0431] Received push event repo: https://github.com/dacofr/hello-appset, revision: main, touchedHead: true 
DEBU[0431] https://github.com/dacofr/hello-appset-copy.git uses repoURL main 
INFO[0431] refresh ApplicationSet argocd/hello-git-appset-copy from webhook 
DEBU[0431] https://github.com/dacofr/hello-appset.git uses repoURL main 
INFO[0431] refresh ApplicationSet argocd/hello-git-appset from webhook 
```

**Expected behavior**

Only one of ApplicationSet should be refreshed.

### PR DESCRIPTION
**Pull request by dacofr**
Fixes #20485 
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [X] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [X] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [X] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by nitishfy — CHANGES_REQUESTED** [review]
Can you fix the failing tests?

**Comment by dacofr** [comment]
I've feeling that e2e test are flapping or "unstable". I just push to trigger CI again and this times 3 over 4 e2e tests are green without changes from my side.
Other PR have same behaviour.

**Comment by nitishfy** [comment]
> I've feeling that e2e test are flapping or "unstable". I just push to trigger CI again and this times 3 over 4 e2e tests are green without changes from my side. Other PR have same behaviour.

I can see some [other PRs](https://github.com/argoproj/argo-cd/pull/20491) having green CI though. Can you verify the output of `make test-local` locally?

**Comment by blakepettersson** [comment]
>I've feeling that e2e test are flapping or "unstable". I just push to trigger CI again and this times 3 over 4 e2e tests are green without changes from my side.

Yeah I've also seen this lately, not sure what the reason for this is 🤔

**Review by blakepettersson — APPROVED** [review]
LGTM

**Comment by dacofr** [comment]
For any action from my side, this time all e2e test are green :smiley: 
@nitishfy would you mind having a look again in PR :pray:

File: applicationset/webhook/webhook_test.go [hunk 1]
```diff
@@ -199,6 +199,7 @@ func TestWebhookHandler(t *testing.T) {
 		t.Run(test.desc, func(t *testing.T) {
 			fc := fake.NewClientBuilder().WithScheme(scheme).WithObjects(
 				fakeAppWithGitGenerator("git-github", namespace, "https://github.com/org/repo"),
+				fakeAppWithGitGenerator("git-github-copy", namespace, "https://github.com/org/repo-copy"),
```
Comment (andrii-korotkov-verkada) [inline applicationset/webhook/webhook_test.go:202]: Do we need to do extra checks below to make sure the webhook is not triggered wrongly, or are these checks already present?

**Review by andrii-korotkov-verkada — APPROVED** [review]

File: applicationset/webhook/webhook_test.go (hunk 1 above)
Comment (dacofr) [reply applicationset/webhook/webhook_test.go:202]: Checks are already existing :+1: 
Here we have the list of expected appset to be refreshed : https://github.com/argoproj/argo-cd/blob/master/applicationset/webhook/webhook_test.go#L66
and here we have a condition to detect if an non expected appset has been refreshed or not : https://github.com/argoproj/argo-cd/blob/master/applicationset/webhook/webhook_test.go#L246

**Comment by dacofr** [comment]
PR up-to-date with master :smiley:

**Review by andrii-korotkov-verkada — APPROVED** [review]

**Review by ishitasequeira — APPROVED** [review]
The change looks good. Thanks @dacofr !!

**Pull request merged**